### PR TITLE
[MRv2] Multi-config sampler warmup for seeded native and greedy paths (#54425, #54455)

### LINKED ISSUE #54425: [Bug]: V2 sampler warmup misses explicit-seed native path when FlashInfer is enabled
### Your current environment

Field deployment:

- 4× NVIDIA RTX PRO 6000 Blackwell Server Edition GPUs (97,887 MiB each)
- NVIDIA driver 595.58.03, CUDA 13.2
- TP4 / DCP4, V2 model runner, FlashInfer sampler enabled
- Full GLM-5.3 checkpoint on a downstream vLLM/B12X image

The fatal field reproduction was on a downstream build, not stock vLLM `main`. I am filing this upstream because the same dispatch invariant is present in upstream commit [`5e71a11e`](https://github.com/vllm-project/vllm/commit/5e71a11eb2b595ede15ecc39c7dddca38e03deb5).

### 🐛 Describe the bug

V2 sampler warmup does not exercise the native sampling path selected by an explicit per-request seed when FlashInfer is enabled. Consequently, the engine can become ready before native top-k/top-p and Gumbel kernels have compiled or made their first temporary allocations. The first seeded API request can then JIT/allocate a materially different path and OOM under tight residual headroom.

The source path at `5e71a11e` is deterministic:

1. [`SamplingParams.for_sampler_warmup()`](https://github.com/vllm-project/vllm/blob/5e71a11eb2b595ede15ecc39c7dddca38e03deb5/vllm/sampling_params.py#L1234-L1253) says it exercises all sampler logic, but supplies no `seed`.
2. [`warmup_kernels()`](https://github.com/vllm-project/vllm/blob/5e71a11eb2b595ede15ecc39c7dddca38e03deb5/vllm/v1/worker/gpu/warmup.py#L257-L283) assigns that unseeded configuration to all warmup requests.
3. The V2 sampler [disables FlashInfer when `any_explicit_seed()` is true](https://github.com/vllm-project/vllm/blob/5e71a11eb2b595ede15ecc39c7dddca38e03deb5/vllm/v1/worker/gpu/sample/sampler.py#L284-L320). With no seed, warmup takes FlashInfer; a seeded request takes native top-k/top-p plus Gumbel.
4. Although current warmup includes a one-request decode step, it remains unseeded. If the native path were selected, a batch below eight rows would [use the PyTorch full-vocabulary sort path](https://github.com/vllm-project/vllm/blob/5e71a11eb2b595ede15ecc39c7dddca38e03deb5/vllm/v1/sample/ops/topk_topp_sampler.py#L352-L388), another path not covered by the unseeded one-request warmup.

A mixed seeded/unseeded warmup batch would not cover both backends: `any_explicit_seed()` disables FlashInfer for the whole sampled batch.

### Reproduction and field evidence

A source-level reproduction does not require an OOM:

1. Enable the V2 model runner and FlashInfer sampling.
2. Instrument `flashinfer_sample`, `apply_top_k_top_p`, and `gumbel_sample`.
3. Run engine warmup. Only the unseeded FlashInfer path is selected.
4. Send a one-request stochastic completion with an explicit `seed`. Native top-k/top-p and Gumbel are selected for the first time after readiness.

The memory impact was observed on a near-capacity field deployment:

- A fixed 4,518,907,904-byte KV allocation per GPU passed deterministic startup and a 32K retrieval gate.
- The first `temperature=1` request with an explicit seed, 1,027 API prompt tokens, and 512 requested output tokens OOMed in top-p sampling and left workers hung with only 3–169 MiB physically free per rank.
- Reducing fixed KV to 3,415,867,392 bytes per GPU returned approximately 1.027 GiB/rank. The same seeded first-use path and subsequent C8 workload passed.
- On an exact-image control, `_gumbel_sample_kernel` first JIT-compiled after API application startup, directly showing that engine warmup had not exercised the runtime path.

The rejected arm's raw fatal trace was not retained, so I cannot make a more specific claim about the final failed allocation size. The request shape, top-p location, post-ready native-path selection, worker hang, and per-rank physical-memory minima were retained. Full field accounting and caveats are recorded [here](https://github.com/malaiwah/glm52-exl3-vast/blob/e0a84f9be9781618db0879d17e986c5a63897eb7/TEST_RESULTS.md#L208-L341).

### Expected behavior

Before reporting readiness, warmup should exercise both supported backend classes that may JIT or allocate differently:

- unseeded FlashInfer sampling;
- explicitly seeded native top-k/top-p plus Gumbel sampling, including a one-request/small-row shape.

### Suggested fix

Run separate sampler warmup passes for the unseeded and explicitly seeded configurations. Preserve the existing one-request decode shape for the seeded pass so the small-row native implementation is covered. A regression test could instrument backend dispatch and assert that FlashInfer, native top-k/top-p, and Gumbel have all run before warmup returns.

Related but not duplicate reports: #23322, #30637, and #33920 cover sampler/KV memory or full-vocabulary-sort OOMs, but not the seed-dependent FlashInfer/native dispatch gap.

### Before submitting a new issue

- [x] I searched existing issues for `for_sampler_warmup`, `any_explicit_seed`, seeded sampler warmup, FlashInfer, and Gumbel warmup.
- [x] I checked the behavior against current upstream source rather than assuming the downstream field build matched upstream.

### LINKED ISSUE #54455: [Bug]: sampler warmup misses the plain-greedy path — spec-decode rejection kernels JIT on the first request
Sibling of #54425 — same function, a second missing dimension. Filing separately because the
missing path and the observable failure are different; I have not touched that issue.

`SamplingParams.for_sampler_warmup()` is documented as "Set parameters to exercise all sampler
logic", but every warmup request gets `temperature=0.9`, so the warmup never takes the
`needs_logits_processing == False` early return in `Sampler.apply_sampling_params()` — the path
a plain `temperature=0` request uses.

With spec decoding that has a visible consequence. The startup warmup *does* reach
`rejection_sample()` (`_run_decode_step` populates `scheduled_spec_decode_tokens`), but it only
ever arrives there in fp32. dtype is a Triton compile key, so
`_compute_local_logits_stats_kernel`, `_rejection_kernel` and `_resample_kernel` are never
compiled for the model dtype and JIT on the first real greedy request:

```
startup (for_sampler_warmup) : needs_logits_processing=True  -> fp32
first temperature=0 request  : needs_logits_processing=False -> bfloat16   <- JIT here
```

`num_logits` is the same on both sides, so shape is held constant; warming the model dtype
alone takes the three `jit_monitor` warnings to zero. Measured first-request penalty on my box:
**+613 ms (52.1%)**. With `--jit-monitor-mode=error` the monitor is configured to raise on that
request (`vllm/utils/jit_monitor.py`), so this is not only a latency issue.

### The general shape (why this and #54425 are the same bug)

`warmup()` enumerates the **batch-shape** space carefully — `decode_steps`
(`vllm/v1/worker/gpu/warmup.py` L391-409 on main `@cacc429f6`) covers all-spec, mixed
spec/non-spec, single-request and no-draft-token variants, and L222-225 budgets for exactly
that (`num_decode_steps = 5 if num_spec_steps > 0 else 3`). But it samples the
**sampling-parameter** space at exactly one point — L280:

```python
sampling_params = SamplingParams.for_sampler_warmup()   # one object, all requests
```

#54425 is one axis of that point (no `seed` → FlashInfer is chosen, the seeded native
top-k/top-p + Gumbel path never runs). This report is another (`temperature=0.9` →
`needs_logits_processing` short-circuit never runs → the greedy dtype is never compiled).
Both are "the docstring says all sampler logic, and one dimension is pinned".

### Suggested fix

Same shape as #54425's: **run more than one sampler warmup configuration**, rather than
special-casing either axis. Concretely, either

1. add a second pass with a plain `SamplingParams()` (greedy) alongside the existing one, or
2. have `for_sampler_warmup()` return a small *list* of configurations and let `warmup()`
   iterate over it the way it already iterates over `decode_steps` (L409) — which would
   close #54425 and this one with the same change, and rides the structure already there.

I have a working local patch for the narrow greedy case (18 lines in
`vllm/v1/worker/gpu/warmup.py`) that takes the JIT warnings to zero, but it clears an internal
sampler buffer for the replay, which I don't think is the right shape to upstream. Happy to
send whichever of (1)/(2) you prefer as a PR.

### Environment

Intel XPU (Arc Pro B70), vLLM main `@cacc429f6` (2026-08-28), spec decode `qwen3_5_mtp`,
Qwen3.6-35B-A3B MXFP4. I carry one unrelated local patch (an mxfp4 `register_fake`);
`sampling_params.py` and `warmup.py` are stock. No CUDA box here to confirm, and as far as I
can tell nothing in this path branches on platform, but I can't verify that claim on hardware.

### PR DESCRIPTION
**Pull request by bandham-manikanta**
### Purpose
Fixes #54425.
Fixes #54455.

### Root Cause
`warmup_kernels()` used a single unseeded stochastic `SamplingParams` for all warmup requests, leaving two dispatch paths uncompiled before serving:

1. **Seeded (#54425)**: Explicit seeds bypass FlashInfer -> native `gumbel_sample` / `apply_top_k_top_p`. First seeded request JIT-compiles these kernels live, risking OOM under tight KV-cache headroom.
2. **Greedy spec-decode (#54455)**: `temperature=0.0` skips logits processing and uses `bfloat16` rejection kernels. These are never compiled during warmup, causing a live Triton JIT spike (+613 ms measured by the issue reporter) on the first greedy request.

### Fix
1. Added `SamplingParams.for_all_sampler_warmup_configs()` returning `[stochastic, seeded(seed=42), greedy(temp=0.0)]`.
2. Each config gets its own isolated `_run_decode_step` pass. Slot `0` is reused when `num_reqs == 1` to stay within batch limits (the CodeRabbit-flagged edge case).
3. Unit tests in `tests/v1/worker/test_gpu_model_runner.py` and `tests/v1/worker/test_gpu_sampler_flags.py` assert all three dispatch paths are exercised.

### Test Plan
`pytest tests/v1/worker/test_gpu_model_runner.py`
`pytest tests/v1/worker/test_gpu_sampler_flags.py`

### Test Results
- **Unit tests pass**: All 3 warmup configs (stochastic, seeded, greedy) confirmed present in the batch passed to the sampler.
- **Cold-start latency (A100)**: First seeded/greedy request completes in ~8 ms -- no JIT spike.

cc @malaiwah @TSUMUGI-XE

---
*Disclaimer: AI assistance was used for unit test setup and documentation for this PR.*

### CODE REVIEW

**Comment by TSUMUGI-XE** [comment]
Tested on Intel XPU. **It fixes it: 3 JIT compilations → 0, and the first-request penalty
drops from +94% to +15%.**

**Setup**: 1× Intel Arc Pro B70, Qwen3.6-35B-A3B MXFP4, spec decode `qwen3_5_mtp`
(`num_speculative_tokens=2`), `cudagraph_mode=FULL_DECODE_ONLY`, `kv-cache-dtype fp8`.
Patch applied as a diff onto `cacc429f6` (all three files applied cleanly). Requests are
`temperature=0` — the path #54455 was about. **No warmup workaround of my own in either arm**,
otherwise the control would not show the effect.

```
                   1st req    reqs 2-4    penalty        JIT warnings
control (no PR)     1.127 s     0.581 s   +93.7% (+545 ms)     3
this PR             0.664 s     0.581 s   +14.5%  (+84 ms)     0
```

The three that disappear are exactly the ones from the report:
`_compute_local_logits_stats_kernel`, `_rejection_kernel`, `_resample_kernel`.

Steady state is identical (0.581 s in both), so no regression on this box.

**One honest note**: the remaining +84 ms is *not* JIT — `jit_monitor` reports zero. So there
is some other first-request cost left over that this PR is not responsible for and that I have
not chased. Worth knowing if someone measures the same thing and wonders why it is not 0%.

**Comment by bandham-manikanta** [comment]
Thanks a lot @TSUMUGI-XE for testing this on Intel XPU!

**Comment by bandham-manikanta** [comment]
Rebased onto `main` and ready for review. Unit tests in `tests/v1/worker/test_gpu_sampler_flags.py` pass (14/14) and verified on A100.

cc @njhill @NickLucche — could someone please trigger the CI run?

File: vllm/v1/worker/gpu/warmup.py [hunk 1]
```diff
@@ -309,7 +309,7 @@ def _alloc_blocks(num_blocks: int) -> list[int]:
             Request(
                 req_ids[i],
                 prompt_token_ids,
-                sampling_params,
+                sampling_params_list[i % len(sampling_params_list)],
```
[the remark this answers is not in the document]
Comment (bandham-manikanta) [reply vllm/v1/worker/gpu/warmup.py:312]: @coderabbitai The batch limit violation for the seeded and greedy sampling paths has been addressed in the latest commit. We now dispatch them via isolated _run_decode_step passes at the end of the warmup to ensure they are exercised even when num_reqs < 3.

**Comment by bandham-manikanta** [comment]
Hi @njhill @NickLucche @WoosukKwon — could one of you kick off CI for this PR? The fix is targeted and self-contained (two new `_run_decode_step` passes for seeded and greedy warmup paths; unit tests included). Happy to address any review feedback.

Related: this is the sampler-space sibling of #55341 (warmup ordering) that @aoshen02 just merged.
